PokemonFactory.sol

```diff
@@ -4,32 +4,112 @@ pragma solidity >=0.7.0 <0.9.0;
     mapping (address => uint) ownerPokemonCount;
 
-     function createPokemon (string memory _name, uint _id) public {
-        pokemons.push(Pokemon(_id, _name));
-        pokemonToOwner[_id] = msg.sender;
-        ownerPokemonCount[msg.sender]++;
+    mapping (uint => string) public pokemonTypes;
+    uint public amountOfTypes; 
+
+    event NewPokemon(uint id, string name, string imageUri, Ability[] abilities, Type[] types, Type[] weaknesses);
+
+    constructor(string[] memory _types) {
+        pokemons.push();
+        amountOfTypes = _types.length;
+
+        for(uint i = 0; i < _types.length; i++) {
+            pokemonTypes[i] = _types[i];
+        }
     }
 
-    function getAllPokemons() public view returns (Pokemon[] memory) {
-      return pokemons;
+     function createPokemon (
+            uint _id, 
+            string memory _name, 
+            string memory _imageUri,
+            string[] memory _skillNames, 
+            string[] memory _skillDescriptions, 
+            uint[] memory _types, 
+            uint[] memory _weaknesses
+        ) public {
+            require(_id > 0, "El id debe ser mayor a 0!");
+            require(pokemonToOwner[_id] == address(0), "Ese id ya esta utilizado");
+            require(bytes(_name).length >= 2, "El nombre debe ser mayor a 2 caracteres!");
+            require(_skillNames.length == _skillDescriptions.length, "Se deben tener la misma cantidad de nombres y descripciones de habilidades");
+            require(_types.length > 0, "Debes tener al menos un tipo");
+            require(_weaknesses.length > 0, "Debes tener al menos una debilidad");
+
+            pokemons.push();
+            uint256 newIndex = pokemons.length - 1;
+            pokemons[newIndex].id = _id;
+            pokemons[newIndex].name = _name;
+            pokemons[newIndex].imageUri = _imageUri;
+
+            for(uint i = 0; i < _skillNames.length; i++){
+                pokemons[newIndex].abilities.push(
+                    Ability({
+                        name: _skillNames[i], 
+                        description: _skillDescriptions[i]
+                        })
+                    );
+            }            
+
+            for(uint i = 0; i < _types.length; i++){
+                require(_types[i] <= amountOfTypes, "Debes mandar un id de tipo valido");
+                pokemons[newIndex].types.push(
+                    Type({
+                        id: i, 
+                        name: pokemonTypes[_types[i]]
+                        })
+                    );
+            }    
+
+            for(uint i = 0; i < _weaknesses.length; i++){
+                require(_types[i] <= amountOfTypes, "Debes mandar un id de tipo valido");
+                pokemons[newIndex].weaknesses.push(
+                    Type({
+                        id: i, 
+                        name: pokemonTypes[_weaknesses[i]]
+                        })
+                    );
+            }    
+
```

Comment:
Me gusta esos For que validan cosas, pero hacen el método de agregar pokemon muy grande, que tal si los encapsulamos en métodos más peques? Y los llamamos acá?